sunkit_image/utils/noise.py

```diff
@@ -1,270 +1,306 @@
-
-        Parameters
-        ----------
-        H : `numpy.ndarray`
-            Input matrix.
-        m : `numpy.ndarray`
-            Rows in convolution matrix.
-        n : `numpy.ndarray`
-            Columns in convolution matrix.
-
-        Returns
-        -------
-        T : `numpy.ndarray`
-            The new convoluted matrix.
-        """
-        s = np.shape(H)
-        T = np.zeros([(m - s[0] + 1) * (n - s[1] + 1), m * n])
-
-        k = 0
-        for i in range((m - s[0] + 1)):
-            for j in range((n - s[1] + 1)):
-                for p in range(s[0]):
-                    T[k, (i + p) * n + j : (i + p) * n + j + 1 + s[1] - 1] = H[p, :]
-                k = k + 1
-        return T
-
-    def weaktexturemask(self):
-        """
-        Calculates the weak texture mask.
-        """
-        if len(self.img.shape) < 3:
-            self.img = np.expand_dims(self.img, 2)
-
-        kh = np.expand_dims(np.transpose(np.vstack(np.array([-0.5, 0, 0.5]))), 2)
-        imgh = correlate(self.img, kh, mode="nearest")
-        imgh = imgh[:, 1 : imgh.shape[1] - 1, :]
-        imgh = imgh * imgh
-
-        kv = np.expand_dims(np.vstack(np.array([-0.5, 0, 0.5])), 1)
-        imgv = correlate(self.img, kv, mode="nearest")
-        imgv = imgv[1 : imgv.shape[0] - 1, :, :]
-        imgv = imgv * imgv
-
-        s = self.img.shape
-        msk = np.zeros_like(self.img)
-
-        for cha in range(s[2]):
-            m = view_as_windows(self.img[:, :, cha], (self.patchsize, self.patchsize))
-            m = np.zeros_like(
-                m.reshape(
-                    np.int(m.size / self.patchsize ** 2), self.patchsize ** 2, order="F"
-                ).transpose()
-            )
-
-            Xh = view_as_windows(imgh[:, :, cha], (self.patchsize, self.patchsize - 2))
-            Xh = Xh.reshape(
-                np.int(Xh.size / ((self.patchsize - 2) * self.patchsize)),
-                ((self.patchsize - 2) * self.patchsize),
-                order="F",
+                break
+
+            cov = (np.asmatrix(X) @ np.asmatrix(X).getH()) / (X.shape[1] - 1)
+            d = np.flip(np.linalg.eig(cov)[0], axis=0)
+            sig2 = d[0]
+
+        nlevel[cha] = np.sqrt(sig2)
+        th[cha] = tau
+        num[cha] = X.shape[1]
+
+    # clean up
+    img = np.squeeze(img)
+
+    return nlevel, th, num
+
+
+def convmtx2(H, m, n):
+    """
+    Specialized 2D convolution matrix generation.
+
+    Parameters
+    ----------
+    H : `numpy.ndarray`
+        Input matrix.
+    m : `numpy.ndarray`
+        Rows in convolution matrix.
+    n : `numpy.ndarray`
+        Columns in convolution matrix.
+
+    Returns
+    -------
+    T : `numpy.ndarray`
+        The new convoluted matrix.
+    """
+    s = np.shape(H)
+    m = int(m)
+    n = int(n)
+    T = np.zeros([(m - s[0] + 1) * (n - s[1] + 1), m * n])
+
+    k = 0
+    for i in range((m - s[0] + 1)):
+        for j in range((n - s[1] + 1)):
```

Comment:
These two values are calculated above for the creation of `T`, could you named them and then used it here too? it will make the code easier to read.
For example:
```python
sub_m = m - s[0] + 1
```
but with a meaningful name.